### PR TITLE
fix(ivy): TestBed.get(Compiler) throws "Error: Runtime compiler is not loaded"

### PR DESCRIPTION
**Pull request by mhevery**
This PR adds support for the `TestBed.get(Compiler)` and unblocks FW-561.

fix(ivy): TestBed.get(Compiler) throws "Error: Runtime compiler is not loaded"

BREAKING CHANGE:

The public API for `DebugNode` was accidentally too broad. This change removes
1. Public constructor. Since `DebugNode` is a way for Angular to communicate information
   on to the developer there is no reason why the developer should ever need to
   Instantiate the `DebugNode`
2. We are also removing `removeChild`, `addChild`, `insertBefore`, and `insertChildAfter`.
   All of these methods are used by Angular to constructor the correct `DebugNode` tree.
   There is no reason why the developer should ever be constructing a `DebugNode` tree
   And these methods should have never been made public. 
3. All properties have been change to `readonly` since `DebugNode` is used by Angular 
   to communicate to developer and there is no reason why these APIs should be writable.

While technically breaking change we don’t expect anyone to be effected by this change.

### CODE REVIEW

File: packages/core/src/debug/debug_node.ts [hunk 1]
```diff
@@ -16,12 +21,26 @@ export class EventListener {
 /**
  * @publicApi
  */
-export class DebugNode {
-  listeners: EventListener[] = [];
-  parent: DebugElement|null = null;
+export interface DebugNode {
```
Comment (pkozlowski-opensource) [inline packages/core/src/debug/debug_node.ts]: This is a breaking change, technically speaking, right? I mean, previously people could import and instantiate `DebugNode` and it was part of the public API.

I don't suspect that many people actually do it but we should still document this as a breaking change

File: packages/core/src/debug/debug_node.ts [hunk 2]
```diff
@@ -146,27 +180,198 @@ function _queryElementChildren(
...
+
+  get injector(): Injector { return getInjector(this.nativeNode); }
+
+  get componentInstance(): any {
+    const nativeElement = this.nativeElement;
+    return nativeElement && getComponent(nativeElement);
+  }
+  get context(): any { throw notImplemented(); }
+
+  get references(): {[key: string]: any;} {
```
Comment (pkozlowski-opensource) [inline packages/core/src/debug/debug_node.ts]: Couldn't we use https://github.com/angular/angular/blob/dc300c5c4172e1f8a6f3f84da6797e427e3626c0/packages/core/src/render3/discovery_utils.ts#L185-L193 here?

File: packages/core/src/debug/debug_node.ts [hunk 3]
```diff
@@ -146,27 +180,198 @@ function _queryElementChildren(
...
+    const lView = context.lViewData;
+    const tView = lView[TVIEW];
+    const tNode = tView.data[context.nodeIndex] as TNode;
+    const properties = {};
+    // TODO: https://angular-team.atlassian.net/browse/FW-681
+    // Missing implementation here...
+    return properties;
+  }
+
+  get attributes(): {[key: string]: string | null;} { throw notImplemented(); }
```
Comment (pkozlowski-opensource) [inline packages/core/src/debug/debug_node.ts]: What would be the strategy for implementing it in the future? Just get static attrs from `TView`? Would we also need info about attr bindings (`attr.foo="expr"`) here? Same applies to classes and styles.

Generally speaking I'm not clear how the current view engine works in this respect...

File: packages/core/src/debug/debug_node.ts [hunk 4]
```diff
@@ -146,27 +180,198 @@ function _queryElementChildren(
...
+    return matches;
+  }
+
+  get listeners(): EventListener[] {
+    // TODO: add real implementation;
+    return [];
+  }
+
+  triggerEventHandler(eventName: string, eventObj: any): void {
+    const event = document.createEvent('MouseEvent');
```
Comment (pkozlowski-opensource) [inline packages/core/src/debug/debug_node.ts]: Uh, this is big change as compared to the current implementation that tries to abstract native event handlers. Not only the `'MouseEvent'` is hard-coded here but I suspect that it will fail for pseudo-events like `window.keydown.ESC`.


Also I'm not sure if the current implementation triggers event handlers on directives.

tl;dr; is that I was trying to take the approach you are taking in this PR and decided agains party for the above reasons. The existing `DebugElement` abstracts native platform while here we are going low-level native. `TestBed` is a public API and this change would be very breaking for many projects I know (ex. ng-bootstrap).

As a very minimum we need to evaluate this breaking change and document if we decide to go down this path.

File: packages/core/src/render3/context_discovery.ts [hunk 5]
```diff
@@ -262,12 +262,6 @@ function findViaDirective(lViewData: LViewData, directiveInstance: {}): number {
   return -1;
 }
 
-function assertDomElement(element: any) {
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/context_discovery.ts]: 👍 good move moving this one :-)

File: packages/core/src/view/services.ts [hunk 6]
```diff
@@ -780,14 +780,14 @@ export class DebugRenderer2 implements Renderer2 {
     const el = this.delegate.selectRootElement(selectorOrNode, preserveContent);
     const debugCtx = getCurrentDebugContext() || (ivyEnabled ? this.createDebugContext(el) : null);
```
Comment (pkozlowski-opensource) [inline packages/core/src/view/services.ts]: I believe that we can remove the part guided by `ivyEnabled` now.

**Review by pkozlowski-opensource — CHANGES_REQUESTED** [review]
@mhevery I've left few comments in the code, the most important being about breaking changes. I was trying to take the same approach in the past but I got scared off the breaking changes with triggering event handlers etc. Basically we would loose the abstraction over native platform and this sounds like "big deal". Let's discuss.

If we decide to go ahead with this approach it is not clear for me what should happen with https://github.com/angular/angular/blob/master/packages/core/src/render3/debug.ts and all infrastructure that plugs it (https://github.com/angular/angular/blob/master/packages/core/testing/src/r3_test_bed.ts#L426-L429)  

Lastly, the CI is red with some tests failing.

File: packages/core/src/debug/debug_node.ts (hunk 1 above)
Comment (mhevery) [reply packages/core/src/debug/debug_node.ts]: @IgorMinar 

From my point of view we have accidentally exposed more API than we intended to. 
- We exposed the constructor
- We exposed the mutation methods 

By changing this from `class` to `interface` I think we are only removing the API which was accidentally exposed and which serves no purpose to the end user.

File: packages/core/src/debug/debug_node.ts (hunk 2 above)
Comment (mhevery) [reply packages/core/src/debug/debug_node.ts]: you are right, fixed

File: packages/core/src/debug/debug_node.ts (hunk 3 above)
Comment (mhevery) [reply packages/core/src/debug/debug_node.ts]: We have all of the information in `TNode` it is just a matter of getting it. I am not sure what the current behavior is either. sounds like we need to do some investigation.

File: packages/core/src/debug/debug_node.ts (hunk 4 above)
Comment (mhevery) [reply packages/core/src/debug/debug_node.ts]: sorry, this is a temporary hack implementation. Proper implementation to be added. I have added comments to document it.

Comment (mhevery) [reply packages/core/src/debug/debug_node.ts]: https://angular-team.atlassian.net/browse/FW-719

File: packages/core/src/view/services.ts (hunk 6 above)
Comment (mhevery) [reply packages/core/src/view/services.ts]: Good catch.

I removed the check

**Comment by mhevery** [comment]
> If we decide to go ahead with this approach it is not clear for me what should happen with https://github.com/angular/angular/blob/master/packages/core/src/render3/debug.ts and all infrastructure that plugs it (https://github.com/angular/angular/blob/master/packages/core/testing/src/r3_test_bed.ts#L426-L429)

Ahh that is an excellent Catch! Yes that code is dead and I have now deleted it. :-)

**Review by pkozlowski-opensource — APPROVED** [review]
@mhevery  if you are positive that we can implement the rest of the `DebugElement` methods this LGTM
Would be probably good if someone else can do another pass on it as we are touching many parts

File: packages/core/src/debug/debug_node.ts (hunk 1 above)
Comment (JoostK) [reply packages/core/src/debug/debug_node.ts]: @mhevery Could you add to the list of breaking changes that `instanceof DebugNode` checks can no longer be done.

**Comment by mhevery** [comment]
http://test/OCL:222839683:BASE:222839704:1543254316995:47c83647

File: packages/core/src/debug/debug_node.ts (hunk 1 above)
Comment (mhevery) [reply packages/core/src/debug/debug_node.ts]: @JoostK will add.

File: packages/common/test/directives/ng_component_outlet_spec.ts [hunk 7]
```diff
@@ -144,22 +144,21 @@ describe('insert/remove', () => {
...
 
-           fixture.detectChanges();
-           expect(fixture.nativeElement).toHaveText('baz');
-         }));
+       fixture.detectChanges();
+       expect(fixture.nativeElement).toHaveText('baz');
+     }));
 
-  fixmeIvy('FW-561: Runtime compiler is not loaded') &&
+  fixmeIvy('destroy is not being called') &&
```
Comment (kara) [inline packages/common/test/directives/ng_component_outlet_spec.ts]: Can you add the Jira ticket number associated with this bug or mark it "unknown"? Otherwise it's hard to tell whether this root cause has been triaged.

File: packages/core/src/render3/assert.ts [hunk 8]
```diff
@@ -83,3 +83,7 @@ function throwError(msg: string): never {
   debugger;  // Left intentionally for better debugger experience.
   throw new Error(`ASSERTION ERROR: ${msg}`);
 }
+
+export function assertDomNode(node: any) {
+  assertEqual(node instanceof Node, true, 'The provided value must be an instance of an DOM Node');
```
Comment (kara) [inline packages/core/src/render3/assert.ts]:
```suggestion
  assertEqual(node instanceof Node, true, 'The provided value must be an instance of a DOM Node');
```

File: packages/core/test/linker/ng_module_integration_spec.ts [hunk 9]
```diff
@@ -284,39 +276,36 @@ function declareTests(config?: {useJit: boolean}) {
...
       afterEach(() => clearModulesForTest());
 
-      fixmeIvy('FW-561: Runtime compiler is not loaded') &&
-          it('should register loaded modules', () => {
-            createModule(SomeModule);
-            const factory = getModuleFactory(token);
-            expect(factory).toBeTruthy();
-            expect(factory.moduleType).toBe(SomeModule);
-          });
+      fixmeIvy('No module with ID myid loaded') && it('should register loaded modules', () => {
```
Comment (kara) [inline packages/core/test/linker/ng_module_integration_spec.ts]: Reason should have Jira ticket or "unknown" here?

**Review by kara — APPROVED** [review]

File: packages/core/src/debug/debug_node.ts (hunk 1 above)
Comment (IgorMinar) [reply packages/core/src/debug/debug_node.ts]: @mhevery it should be fine to change this behavior, but we need to document it.

File: packages/common/test/directives/ng_component_outlet_spec.ts (hunk 7 above)
Comment (mhevery) [reply packages/common/test/directives/ng_component_outlet_spec.ts]: destroy on NgModuleRef is not being called

File: packages/core/src/render3/assert.ts (hunk 8 above)
Comment (mhevery) [reply packages/core/src/render3/assert.ts]: fixed

File: packages/core/test/linker/ng_module_integration_spec.ts (hunk 9 above)
Comment (mhevery) [reply packages/core/test/linker/ng_module_integration_spec.ts]: https://angular-team.atlassian.net/browse/FW-740

File: packages/core/src/metadata/directives.ts [hunk 10]
```diff
@@ -77,6 +77,9 @@ export interface DirectiveDecorator {
   new (obj: Directive): Directive;
 }
 
+/**
+ * @publicApi
```
Comment (IgorMinar) [inline packages/core/src/metadata/directives.ts]: why do we declare and export directive interface twice? that looks really fishy. Can we remove one?

File: tools/public_api_guard/core/testing.d.ts [hunk 11]
```diff
@@ -47,6 +47,52 @@ export declare type MetadataOverride<T> = {
 
 export declare function resetFakeAsyncZone(): void;
 
+export interface TestBed {
```
Comment (IgorMinar) [inline tools/public_api_guard/core/testing.d.ts]: I can't believe that we missed having all of these interfaces in the guard. Good catch!!!

**Review by IgorMinar — APPROVED** [review]
The rest looks good

File: tools/public_api_guard/core/testing.d.ts (hunk 11 above)
Comment (mhevery) [reply tools/public_api_guard/core/testing.d.ts]: fixed

**Comment by mhevery** [comment]
http://test/OCL:222881368:BASE:222938072:1543299184520:907394cc

**Comment by pkozlowski-opensource** [comment]
@mhevery looks like CI is red (the golden symbols file need updating etc.)

**Comment by mhevery** [comment]
http://test/OCL:222839683:BASE:223003683:1543339823046:c3644f5e
